blueocean-personalization/src/main/js/model/FavoriteStore.js

```diff
@@ -0,0 +1,128 @@
+import { observable, action } from 'mobx';
+import { UrlConfig, Fetch, AppConfig, User, capabilityAugmenter, ToastService, i18nTranslator } from '@jenkins-cd/blueocean-core-js';
+import { cleanSlashes } from '../util/UrlUtils';
+import { FavoritesSortHelper } from '../util/SortUtils';
+import { blueocean } from '@jenkins-cd/blueocean-core-js/dist/js/scopes';
+
+const t = i18nTranslator('blueocean-personalization');
+const sortHelper = new FavoritesSortHelper();
+
+const defaultBranchName = 'master';
+
+function getItemName(item) {
+    const fullName = item.fullName;
+    if (fullName.indexOf('/') < 0) {
+        return fullName + '/' + defaultBranchName;
+    }
+    return fullName;
+}
+
+class FavoriteStore {
+    @observable _favoritesNames = blueocean.favoritesList;
+    @observable _favorites = [];
+    _fetched = false;
+
+    @action
+    _setFavoritesNames(_favoritesNames) {
+        this._favoritesNames = _favoritesNames;
+    }
+
+    @action
+    _setFavorites(_favorites) {
+        this._favorites = _favorites;
+    }
+
+    @action
+    setFavorite(item, favorite) {
+        if (favorite === this.isFavorite(item)) {
+            return; // nothing to do
+        }
+
+        const baseUrl = UrlConfig.getJenkinsRootURL();
+        const url = cleanSlashes(`${baseUrl}${item._links.self.href}/favorite`);
+
+        const fetchOptions = {
+            method: 'PUT',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify({ favorite: favorite }),
+        };
+
+        this.fetch(url, fetchOptions)
+        .then(() => {
+            const { _favoritesNames: f } = this;
+            const itemName = getItemName(item);
+            if (favorite) {
+                this._setFavoritesNames([...f, itemName]);
+            } else {
+                const idx = f.indexOf(itemName);
+                this._setFavoritesNames([...f.slice(0, idx), ...f.slice(idx + 1)]);
```

Comment:
This might read slightly better as an array.filter

Reply:
Faster, too. But not a big deal.